src/mongo/db/telemetry/telemetry_thread_base.h

```diff
@@ -0,0 +1,136 @@
+/*======
+This file is part of Percona Server for MongoDB.
+
+Copyright (C) 2024-present Percona and/or its affiliates. All rights reserved.
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the Server Side Public License, version 1,
+    as published by MongoDB, Inc.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    Server Side Public License for more details.
+
+    You should have received a copy of the Server Side Public License
+    along with this program. If not, see
+    <http://www.mongodb.com/licensing/server-side-public-license>.
+
+    As a special exception, the copyright holders give permission to link the
+    code of portions of this program with the OpenSSL library under certain
+    conditions as described in each individual source file and distribute
+    linked combinations including the program with the OpenSSL library. You
+    must comply with the Server Side Public License in all respects for
+    all of the code used other than as permitted herein. If you modify file(s)
+    with this exception, you may extend this exception to your version of the
+    file(s), but you are not obligated to do so. If you do not wish to do so,
+    delete this exception statement from your version. If you delete this
+    exception statement from all source files in the program, then also delete
+    it in the license file.
+======= */
+
+#pragma once
+
+#include <boost/filesystem.hpp>  // IWYU pragma: keep
+#include <memory>
+#include <string>
+
+#include "mongo/base/status.h"
+#include "mongo/base/string_data.h"
+#include "mongo/bson/bsonobj.h"
+#include "mongo/bson/bsonobjbuilder.h"
+#include "mongo/bson/oid.h"
+#include "mongo/db/operation_context.h"
+#include "mongo/platform/atomic_word.h"
+#include "mongo/platform/mutex.h"
+#include "mongo/stdx/condition_variable.h"
+#include "mongo/util/background.h"
+#include "mongo/util/time_support.h"
+
+namespace mongo {
+
+class ServiceContext;
+
+class TelemetryThreadBase : public BackgroundJob {
+public:
+    // Pointer to the function to create subclass instance
+    static std::unique_ptr<TelemetryThreadBase> (*create)();
```

Comment:
As far as I remember I tried to do that but failed because `telemetry_thread_d.cpp` and `telemetry_thread_base.cpp` are in different libraries and in case of dynamic linking the linker cannot resolve all symbols.

Reply:
I'll think about another solution that would bring error about potentially uninitialized function pointer.